controllers/products_test.go

```diff
@@ -170,129 +271,38 @@ func TestCreateProduct(t *testing.T) {
-		t.Errorf("Failed to create user uuid: %s", err)
-		return
-	}
-	users[userID] = 0
-
-	// Execute test
-	if err := validateUsers(users); err != nil {
-		t.Errorf("\nTest returned:\n %+v\nExpected:\n %+v", err, nil)
-		return
-	}
-}
-
-func TestValidateUsers_EmptyUsersList(t *testing.T) {
-	// Create test data
-	users, privileges := createTestUsersData()
-	mysqldb.Functions = DBFunctionInterfaceMock{
-		privileges: privileges,
-	}
-
-	// Execute test
-	if err := validateUsers(users); err == nil || (err != nil && err != ErrEmptyUsersList) {
-		t.Errorf("\nTest returned:\n %+v\nExpected:\n %+v", err, ErrEmptyUsersList)
-		return
-	}
-}
-
-func TestValidateUsers_NilUserList(t *testing.T) {
-	// Create test data
-	var users models.ProductUsers
-
-	// Execute test
-	if err := validateUsers(users); err == nil || (err != nil && err != ErrEmptyUsersList) {
-		t.Errorf("\nTest returned:\n %+v\nExpected:\n %+v", err, ErrEmptyUsersList)
-		return
-	}
-}
-
-func TestValidateUsers_NoOwner(t *testing.T) {
-	// Create test data
-	users, privileges := createTestUsersData()
-	mysqldb.Functions = DBFunctionInterfaceMock{
-		privileges: privileges,
-	}
-	userID, err := uuid.NewUUID()
-	if err != nil {
-		t.Errorf("Failed to create user uuid: %s", err)
-		return
-	}
-	users[userID] = 1
-
-	userID, err = uuid.NewUUID()
-	if err != nil {
-		t.Errorf("Failed to create user uuid: %s", err)
-		return
-	}
-	users[userID] = 1
-
-	// Execute test
-	if err := validateUsers(users); err == nil || (err != nil && err != ErrInvalidOwnerCount) {
-		t.Errorf("\nTest returned:\n %+v\nExpected:\n %+v", err, ErrInvalidOwnerCount)
-		return
-	}
-}
-
-func TestValidateUsers_MultipleOwners(t *testing.T) {
-	// Create test data
-	users, privileges := createTestUsersData()
-	mysqldb.Functions = DBFunctionInterfaceMock{
-		privileges: privileges,
-	}
-	userID, err := uuid.NewUUID()
-	if err != nil {
-		t.Errorf("Failed to create user uuid: %s", err)
-		return
-	}
-	users[userID] = 0
-
-	userID, err = uuid.NewUUID()
-	if err != nil {
-		t.Errorf("Failed to create user uuid: %s", err)
+		t.Errorf("Failed to create test data: %s", err)
 		return
 	}
-	users[userID] = 0
 
-	// Execute test
-	if err := validateUsers(users); err == nil || (err != nil && err != ErrInvalidOwnerCount) {
-		t.Errorf("\nTest returned:\n %+v\nExpected:\n %+v", err, ErrInvalidOwnerCount)
-		return
-	}
-}
-
-func TestValidateUsers_InvalidPrivilege(t *testing.T) {
-	// Create test data
-	users, privileges := createTestUsersData()
-	mysqldb.Functions = DBFunctionInterfaceMock{
-		privileges: privileges,
-	}
-	userIDInvalid, err := uuid.NewUUID()
-	if err != nil {
-		t.Errorf("Failed to create user uuid: %s", err)
-		return
-	}
-	users[userIDInvalid] = 2
+	// Run tests
+	for _, testCaseString := range dataSet.OrderedList {
+		testCaseString := testCaseString
+		t.Run(testCaseString, func(t *testing.T) {
+			testCase := dataSet.TestDataSet[testCaseString]
+			var expectedError error
+			if testCase.Expected != nil {
```

Comment:
How do you test the happy paths if your expected value is an error (or nil)?
If I were you, I probably create an expected data and expected error for every test case and one of them will be null, but the other holds the real expectations.

Reply:
You are right, and many cases have expected data as well. Unfortunately, the coverage is not full. Also I am starting to get a better understanding how to do better testing in golang.
In this specific case, you've made the comment. If the user is valid no error returns, so that is the happy path. In any other case a certain error shall return.

Reply:
Suppose I can flick a bool in the return value, but I just feel it is a boiler plate value, if the error gives back all the data I need. (even though customs may suggest to set a boolean as well)